### PR TITLE
Add WLP-InstantOn-Enabled header support

### PR DESCRIPTION
**Pull request by tjwatson**

### CODE REVIEW

**Comment by tjwatson** [comment]
#build

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java [hunk 1]
```diff
@@ -503,14 +372,26 @@ private void checkSupportedFeatures() {
             // shortcut if all features area allowed
             return;
         }
+        BundleContext context = cc.getBundleContext();
+        if (context == null) {
+            return;
```
Comment (dazavala) [inline dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:377]: When would this code lack a bundle context?  During a unit test, perhaps?

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java [hunk 2]
```diff
@@ -93,139 +94,7 @@
...
-                               "mpFaultTolerance-4.0",
-                               "mpHealth-4.0",
-                               "mpJwt-2.1",
-                               "mpMetrics-5.0",
-                               "mpOpenAPI-3.1",
-                               "mpRestClient-3.0",
-                               "mpTelemetry-1.0",
-                               "pages-3.1",
-                               "persistence-3.1",
-                               "persistenceContainer-3.1",
```
Comment (dazavala) [inline dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:216]: This solution does not modify the feature for descriptor `xmlBinding-4.0` for WP10.  I assume we only care about public features w.r.t. `WLP-InstantOn-Enabled`.  E.G. `io.openliberty.xmlBinding.internal-4.0` is now enabled by persistenceContainer-3.1 for WP10, whereas `persistenceContainer-3.0` enables public feature `xmlBinding-3.0` for WP91.

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java [hunk 3]
```diff
@@ -93,139 +94,7 @@
...
-                               "restfulWSClient-3.1",
-                               "servlet-6.0",
-                               "ssl-1.0",
-                               "transportSecurity-1.0",
-                               "webProfile-10.0",
-                               "websocket-2.1"
-        };
-        Set<String> result = new HashSet<>();
-        result.addAll(Arrays.asList(supported));
-        SUPPORTED_FEATURES = Collections.unmodifiableSet(result);
```
Comment (dazavala) [inline dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:227]: All descriptors for public features in WP80/91/60 and MP41/50/60, including the convenience features, contain header `WLP-InstantOn-Enabled: true` as intended. I don't find any unexpected changes to these descriptors; nor do I find any changes to descriptors outside this feature set.

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java [hunk 4]
```diff
@@ -503,14 +372,26 @@ private void checkSupportedFeatures() {
...
                 for (ServiceReference<?> feature : features) {
-                    Object featureName = feature.getProperty("ibm.featureName");
-                    if (!SUPPORTED_FEATURES.contains(featureName) && !allowedFeatures.contains(featureName)) {
-                        unsupported.add(featureName);
+                    LibertyFeature libertyFeature = (LibertyFeature) context.getService(feature);
+                    if (libertyFeature != null) {
+                        try {
+                            String instantonEnabled = libertyFeature.getHeader(INSTANTON_ENABLED_HEADER);
+                            Object featureName = feature.getProperty("ibm.featureName");
+                            if (!(Boolean.parseBoolean(instantonEnabled) || allowedFeatures.contains(featureName))) {
```
Comment (dazavala) [inline dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:389]: Flashed back to DeMorgan's law for this one. The revised logic is identical and LGTM.

File: dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/io.openliberty.mpOpenTracing-3.0.feature [hunk 5]
```diff
@@ -36,4 +36,4 @@ IBM-API-Package: \
 kind=ga
 edition=core
 WLP-Activation-Type: parallel
```
Comment (dazavala) [inline dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/io.openliberty.mpOpenTracing-3.0.feature:38]: Per issue #19980 feature `mpOpenTracing` was removed from MP60 and `mpTelemetry-1.0` was added.  If this is a direct feature replacement (i.e. a rename), should the `mpTelemetry-3.0` feature descriptor declare header `superseded-by`?

**Review by dazavala — DISMISSED** [review]
Approved as I found no necessary changes.  There are a few comments and questions posted against the source changes.

FYI, I discovered link [MicroProfile](https://openliberty.io/docs/latest/reference/feature/micropProfile.html) in the openliberty.io feature overview topic does not work.  The URL has an obvious typo.  David Mueller is handling it.

Lastly, I investigated the `concurrent-1.0/2.0` feature dependencies w.r.t. `mpFaultTolerance`.  Your solution LGTM.

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java (hunk 1 above)
Comment (tjwatson) [reply dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:377]: If the bundle is stopping.  It is a common NPE that happens on shutdown, so I usually always put this safeguard.

File: dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java (hunk 2 above)
Comment (tjwatson) [reply dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:216]: Correct, we only check public features here.

File: dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/io.openliberty.mpOpenTracing-3.0.feature (hunk 5 above)
Comment (tjwatson) [reply dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/io.openliberty.mpOpenTracing-3.0.feature:38]: That would be a question for @donbourne.  IIRC the two are separate technologies, but MP6 is including mpTelemetry in the profile while mpOpenTracing is remaining as an optional standalone specification.

**Comment by tjwatson** [comment]
I only rebased my branch because it was pretty far behind.  Going to do another build before merge:

#build

**Review by dazavala — APPROVED** [review]
Approved.  Commits look that similar to previous review.

**Comment by tjwatson** [comment]
#libby

**Pull request merged**